indico/modules/admin/notices.py

```diff
@@ -0,0 +1,159 @@
+# This file is part of Indico.
+# Copyright (C) 2002 - 2021 CERN
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+
+import dataclasses
+import operator
+import platform
+import re
+import typing as t
+from enum import Enum
+
+import requests
+import yaml
+from celery.schedules import crontab
+from flask import session
+from markupsafe import Markup
+from marshmallow import EXCLUDE, ValidationError, fields, post_load
+from marshmallow_dataclass import dataclass
+from packaging.version import InvalidVersion, Version
+
+import indico
+from indico.core.cache import make_scoped_cache
+from indico.core.celery import celery
+from indico.core.config import config
+from indico.core.db import db
+from indico.core.db.sqlalchemy.util.queries import get_postgres_version
+from indico.core.logger import Logger
+from indico.core.marshmallow import mm
+from indico.util.caching import memoize_request
+from indico.util.string import render_markdown
+from indico.web.flask.templating import template_hook
+
+
+logger = Logger.get('notices')
+notices_cache = make_scoped_cache('notices')
+
+
+class NoticeSeverity(str, Enum):
+    highlight = 'highlight'
+    warning = 'warning'
+    error = 'error'
+
+
+@dataclass(frozen=True)
+class SystemNoticeCriteria:
+    python_version: t.Optional[str] = None
+    postgres_version: t.Optional[str] = None
+    indico_version: t.Optional[str] = None
+
+
+@dataclass(frozen=True)
+class SystemNotice:
+    class Meta:
+        unknown = EXCLUDE
+
+    id: str
+    message: str
+    when: SystemNoticeCriteria
+    severity: NoticeSeverity
+    announcement_bar: bool = False
+
+    def evaluate(self, versions):
+        ops = {
+            '==': operator.eq,
+            '!=': operator.ne,
+            '<': operator.lt,
+            '<=': operator.le,
+            '>': operator.gt,
+            '>=': operator.ge,
+        }
+        for key, value in dataclasses.asdict(self.when).items():
+            if value is None:
+                continue
+            if not (match := re.match(fr'^(?P<op>{"|".join(map(re.escape, ops))})\s*(?P<val>[0-9]\S*)$', value)):
+                logger.warning('Invalid criterion in %s: %s %s', self.id, key, value)
+                return False
+            current_version = versions[key]
```

Comment:
Can it happen that there is a criteria in `SystemNoticeCriteria` which isn't in the yaml config?

Reply:
Yes, that happens all the time, because only criteria that are actually used are in the yaml file.
But `versions` is the dict from `_get_versions` and the dataclass cannot contain keys that aren't there unless someone adds a new field to it, but not return a value for it from that function.